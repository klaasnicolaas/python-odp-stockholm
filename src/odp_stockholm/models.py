"""Asynchronous Python client for ODP Stockholm."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import UTC, datetime
from typing import Any


@dataclass
class DisabledParking:
    """Object representing an DisabledParking model response from the API."""

    location_id: int
    location_type: str

    number: int
    street: str
    address: str
    district: str
    parking_rate: str
    parking_rules: str

    valid_from: datetime
    valid_to: datetime | None
    coordinates: list[float]

    @classmethod
    def from_json(cls: type[DisabledParking], data: dict[str, Any]) -> DisabledParking:
        """Return DisabledParking object from a dictionary.

        Args:
        ----
            data: The JSON data from the API.

        Returns:
        -------
            An DisabledParking object.

        """
        attr = data["properties"]
        return cls(
            location_id=attr["FID"],
            location_type=attr["VF_PLATS_TYP"],
            number=attr["EXTENT_NO"],
            street=attr["STREET_NAME"],
            address=attr["ADDRESS"],
            district=attr["CITY_DISTRICT"],
            parking_rate=attr["PARKING_RATE"],
            parking_rules=attr["RDT_URL"],
            valid_from=datetime.strptime(
                attr.get("VALID_FROM"),
                "%Y-%m-%dT%H:%M:%SZ",
            ).replace(tzinfo=UTC),
            valid_to=datetime.strptime(
                attr.get("VALID_TO"),
                "%Y-%m-%dT%H:%M:%SZ",
            ).replace(tzinfo=UTC)
            if attr.get("VALID_TO")
            else None,
            coordinates=data["geometry"]["coordinates"],
        )
