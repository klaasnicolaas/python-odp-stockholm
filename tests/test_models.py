"""Test the models for ODP Stockholm."""

from __future__ import annotations

from aiohttp import ClientSession
from aresponses import ResponsesMockServer

from odp_stockholm import DisabledParking, ParkingStockholm

from . import load_fixtures


async def test_all_disabled_parkings(aresponses: ResponsesMockServer) -> None:
    """Test all disabled parkings function."""
    aresponses.add(
        "openparking.stockholm.se",
        "/LTF-Tolken/v1/prorelsehindrad/all",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixtures("disabled_parking.json"),
        ),
    )
    async with ClientSession() as session:
        client = ParkingStockholm(api_key="test", session=session)
        parking: list[DisabledParking] = await client.disabled_parkings()
        assert parking is not None
