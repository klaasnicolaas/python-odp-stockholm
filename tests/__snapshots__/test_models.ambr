# serializer version: 1
# name: test_all_disabled_parkings
  list([
    DisabledParking(location_id=6205721, location_type='Reserverad p-plats rörelsehindrad', number=1, street='Nathorstvägen', address='Nathorstvägen 21', district='Hammarbyhöjden', parking_rate='taxa 4: 10 kr/tim vardagar 7-19, dag före helgdag 11-17, Boende: 500 kr/månad eller 35 kr/dygn', parking_rules='https://rdt.transportstyrelsen.se/rdt/AF06_View.aspx?BeslutsMyndighetKod=0180&BeslutadAr=2016&LopNr=02649', valid_from=datetime.datetime(2016, 11, 6, 22, 0, tzinfo=<UTC>), valid_to=None, coordinates=[[18.089936, 59.296964], [18.089853, 59.297003]]),
    DisabledParking(location_id=6205722, location_type='Reserverad p-plats rörelsehindrad', number=1, street='Olaus Magnus Väg', address='Olaus Magnus Väg 44A', district='Hammarbyhöjden', parking_rate='taxa 4: 10 kr/tim vardagar 7-19, dag före helgdag 11-17, Boende: 500 kr/månad eller 35 kr/dygn', parking_rules='https://rdt.transportstyrelsen.se/rdt/AF06_View.aspx?BeslutsMyndighetKod=0180&BeslutadAr=2016&LopNr=02653', valid_from=datetime.datetime(2016, 11, 6, 22, 0, tzinfo=<UTC>), valid_to=None, coordinates=[[18.093412, 59.296867], [18.093512, 59.296851]]),
    DisabledParking(location_id=6205723, location_type='Reserverad p-plats rörelsehindrad', number=1, street='Hoburgsstigen', address='Hoburgsstigen 5A - 7', district='Traneberg', parking_rate='taxa 4: 10 kr/tim vardagar 7-19, dag före helgdag 11-17, Boende: 500 kr/månad eller 35 kr/dygn', parking_rules='https://rdt.transportstyrelsen.se/rdt/AF06_View.aspx?BeslutsMyndighetKod=0180&BeslutadAr=2016&LopNr=02656', valid_from=datetime.datetime(2016, 9, 29, 20, 0, tzinfo=<UTC>), valid_to=None, coordinates=[[17.9792, 59.340189], [17.979235, 59.340195], [17.97924, 59.340195], [17.979246, 59.340195], [17.979344, 59.3402], [17.979351, 59.3402], [17.979357, 59.340199], [17.979364, 59.340199], [17.97937, 59.340198], [17.979416, 59.340188]]),
    DisabledParking(location_id=6205724, location_type='Reserverad p-plats rörelsehindrad', number=1, street='Paternostervägen', address='Paternostervägen 52', district='Hammarbyhöjden', parking_rate='taxa 4: 10 kr/tim vardagar 7-19, dag före helgdag 11-17, Boende: 500 kr/månad eller 35 kr/dygn', parking_rules='https://rdt.transportstyrelsen.se/rdt/AF06_View.aspx?BeslutsMyndighetKod=0180&BeslutadAr=2016&LopNr=02661', valid_from=datetime.datetime(2016, 11, 6, 22, 0, tzinfo=<UTC>), valid_to=None, coordinates=[[18.101559, 59.291045], [18.101602, 59.291034], [18.101655, 59.291024]]),
    DisabledParking(location_id=6205725, location_type='Reserverad p-plats rörelsehindrad', number=1, street='Sandfjärdsgatan', address='Sandfjärdsgatan 26B', district='Årsta', parking_rate='taxa 4: 10 kr/tim vardagar 7-19, dag före helgdag 11-17, Boende: 500 kr/månad eller 35 kr/dygn', parking_rules='https://rdt.transportstyrelsen.se/rdt/AF06_View.aspx?BeslutsMyndighetKod=0180&BeslutadAr=2016&LopNr=02819', valid_from=datetime.datetime(2016, 12, 14, 22, 0, tzinfo=<UTC>), valid_to=None, coordinates=[[18.047509, 59.294253], [18.047589, 59.294233], [18.047683, 59.294209]]),
  ])
# ---
